CMakeLists.txt

```diff
@@ -971,6 +971,12 @@ endforeach()
 
 if(CONFIG_KERNEL_WHOLE_ARCHIVE)
   set(WHOLE_ARCHIVE_LIBS ${ZEPHYR_LIBS_PROPERTY} kernel)
+  if(CONFIG_LLEXT)
+    # This library contains weak symbols for all z_impl_* syscalls, and must be
+    # placed after all the actual syscall implementations in the link order to
+    # prevent the weak symbols from overriding the actual implementations.
+    list(APPEND WHOLE_ARCHIVE_LIBS syscall_export_llext)
```

Comment:
we should not have this deep knowledge of `syscall_export_llext` library at this location.

In fact, this commit 321e395a8f4e79ea74b305c248c2f5aec96a4202 should never have been approved in the form it was proposed, but seems I was removed as assignee on the PR (and probably also too busy with HWMv2 at the time).

This change exactly shows the problem with having a `WHOLE_ARCHIVE_LIBS` in its current form.

In general only Zephyr libs should go into the whole-archive and vanilla CMake static will be outside the whole-archive and can be ensured to be picked up when `zephyr_interface` is linked against them.
However, in this specific case, because the CMake static libs has extra dependencies, then that is not possible in this case.

Let me take a closer look of both the needed feature in 321e395a8f4e79ea74b305c248c2f5aec96a4202 and the requirement in this PR and see if I can come up with something cleaner.

Reply:
Both the `WHOLE_ARCHIVE_LIBS` and this PR deal with the issue of allowing extensions to perform syscalls. The most important requirement is that **all _defined_ syscalls need to be preserved** even if unused by the kernel.

We don't strictly need undefined syscalls to have a NULL `_impl` symbol exported, but having everything whole-archive'd implies that _all_ boilerplate is compiled so every symbol must be present at final link time.

Thanks for looking into this, any advice is appreciated.